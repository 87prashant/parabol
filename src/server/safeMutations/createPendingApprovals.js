import getRethink from 'server/database/rethinkDriver'
import {PENDING} from 'server/utils/serverConstants'
import shortid from 'shortid'
import {BILLING_LEADER, REQUEST_NEW_USER} from 'universal/utils/constants'
import makeNewSoftTeamMembers from 'server/safeMutations/makeNewSoftTeamMembers'

export default async function createPendingApprovals (outOfOrgEmails, inviteSender, dataLoader) {
  if (outOfOrgEmails.length === 0) {
    return {
      requestNotifications: [],
      orgApprovalIds: [],
      billingLeaderUserIds: [],
      newSoftTeamMembers: []
    }
  }
  const {orgId, teamId, userId} = inviteSender || {}
  const r = getRethink()
  const now = new Date()
  // add a notification to the Billing Leaders
  const {userIds, inviter} = await r({
    userIds: r
      .table('User')
      .getAll(orgId, {index: 'userOrgs'})
      .filter((user) =>
        user('userOrgs').contains((userOrg) =>
          userOrg('id')
            .eq(orgId)
            .and(userOrg('role').eq(BILLING_LEADER))
        )
      )('id')
      .coerceTo('array'),
    inviter: r
      .table('User')
      .get(userId)
      .pluck('preferredName', 'id')
  })
  const requestNotifications = outOfOrgEmails.map((inviteeEmail) => ({
    id: shortid.generate(),
    type: REQUEST_NEW_USER,
    startAt: now,
    orgId,
    userIds,
    inviterUserId: inviter.id,
    inviteeEmail,
    teamId
  }))

  const pendingApprovals = outOfOrgEmails.map((inviteeEmail) => ({
    id: shortid.generate(),
    createdAt: now,
    email: inviteeEmail,
    isActive: true,
    orgId,
    status: PENDING,
    teamId,
    updatedAt: now
  }))

  // send a new notification to each Billing Leader concerning each out-of-org invitee
  await r({
    requestNotifications: r.table('Notification').insert(requestNotifications),
    approvals: r.table('OrgApproval').insert(pendingApprovals)
  })

  const newSoftTeamMembers = await makeNewSoftTeamMembers(outOfOrgEmails, teamId, dataLoader)

  return {
    requestNotifications,
    newSoftTeamMembers,
    orgApprovalIds: pendingApprovals.map(({id}) => id),
    billingLeaderUserIds: userIds
  }
}
