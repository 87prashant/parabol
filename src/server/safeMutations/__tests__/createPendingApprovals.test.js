import DynamicSerializer from 'dynamic-serializer';
import MockDate from 'mockdate';
import MockDB from 'server/__tests__/setup/MockDB';
import {__now} from 'server/__tests__/setup/mockTimes';
import fetchAndSerialize from 'server/__tests__/utils/fetchAndSerialize';
import newInvitee from 'server/__tests__/utils/newInvitee';
import getRethink from 'server/database/rethinkDriver';
import createPendingApprovals from 'server/safeMutations/createPendingApprovals';

MockDate.set(__now);
console.error = jest.fn();

describe('createPendingApprovals', () => {
  test('returns an empty array if there are no people to approve', async () => {
    // TEST
    const res = await createPendingApprovals([]);

    // VERIFY
    expect(res).toEqual([]);
  });
  test('creates approvals with org leader notifications, returns an object full of the notifications', async () => {
    // SETUP
    const r = getRethink();
    const mockDB = new MockDB();
    const dynamicSerializer = new DynamicSerializer();
    const invitee = newInvitee();
    await mockDB.init();
    const firstUser = mockDB.db.user[0];
    const inviter = {
      userId: firstUser.id,
      teamName: mockDB.context.team.name,
      teamId: mockDB.context.team.id,
      orgId: mockDB.context.organization.id
    };

    // TEST
    const res = await createPendingApprovals([invitee], inviter);

    // VERIFY
    const db = await fetchAndSerialize({
      orgApproval: r.table('OrgApproval').getAll(invitee.email, {index: 'email'}).orderBy('email'),
      notification: r.table('Notification').getAll(inviter.orgId, {index: 'orgId'}).orderBy('userIds')
    }, dynamicSerializer);
    expect(db).toMatchSnapshot();
    expect(res[inviter.userId].length).toEqual(1);
  });
});
