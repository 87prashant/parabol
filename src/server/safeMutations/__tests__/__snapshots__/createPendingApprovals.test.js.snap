// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createPendingApprovals creates approvals with org leader notifications, returns an object full of the notifications 1`] = `
Object {
  "notification": Array [
    Object {
      "id": 0,
      "inviteeEmail": 1,
      "inviterName": "Captain America",
      "inviterUserId": 2,
      "orgId": 3,
      "startAt": 2017-01-01T08:00:00.000Z,
      "teamId": 4,
      "teamName": "Team America",
      "type": "REQUEST_NEW_USER",
      "userIds": Array [
        2,
      ],
    },
  ],
  "orgApproval": Array [],
}
`;
